src/v/cluster/types.h

```diff
@@ -465,11 +465,12 @@ struct fetch_tx_reply
 
 struct begin_tx_request
   : serde::
-      envelope<begin_tx_request, serde::version<0>, serde::compat_version<0>> {
+      envelope<begin_tx_request, serde::version<1>, serde::compat_version<0>> {
     model::ntp ntp;
     model::producer_identity pid;
     model::tx_seq tx_seq;
     std::chrono::milliseconds transaction_timeout_ms{};
+    model::partition_id tm_partition;
```

Comment:
fixed

Reply:
Sorry for pulling you with @dotnwat in different directions maybe this serde part I don't understand but I worry that if `rm_stm` running new code (begin_tx_request with tm_partition) but the tx coordinator runs old code then tx coordinator may invoke begin_tx without setting tm_partition so on the receiver's part it will be uninialized and I suspect it will lead to undefined behavior.

When the caller doesn't pass tm_partition we want it to be initialized with `tm_partition{0}`. Here it isn't obvious what happens. Noah, can you please chime in?

Reply:
I think that best solution there would be `tm_partition{0}` as default initializer for compatibility and adding tm_partition to all constructors.
So we will always provide tm_partition to constructor and we will be able to correctly serialize previous version of request

Reply:
Hmm, I'm not entirely sure I understand the question being asked. The code for this comment is out-dated i think.

The default initialization value for named_type<integers> is not necessarily 0, it's type<T>::min().

Reply:
[there](https://github.com/redpanda-data/redpanda/pull/9362#discussion_r1160332677) you suggested to pass `tm_partition(0)` via constructor but old version doesn't know anything about `tm_partition` so it passes nothing, the request goes over wire to a new node which deserialize it and we have `tm_partition` uninitialized while we want it to be tm_partition(0).

Reply:
the comment had nothing to do with serialization. i was suggesting that we avoid encoding, what seemed like important bits related to backwards compatibility, in the set of constructor overloads and force the callers that "pass nothing" to actually pass "tm_partition(0)" and be explicit about things.

but it's a minor nit you can ignore.